Update for Hashcat v3.00+

nucleare opened the pull request:
As of Hashcat v3.00 _**direct**_ use of `-n` and `-u` would [no longer be supported ](https://hashcat.net/forum/thread-5551-post-29892.html#pid29892) however with some exception, the options for `-n` & `-u` are still recognized, depending on the release version. On the contrary, Hashcat no longer recognizes the command line options for `--gpu-loops` and `--gpu-accel` but will accept `-n` and `-u` following a notice that the user would need to use '--force' to proceed with their "hand-tuned settings". 
Though it's a bit sloppy _(due to lack of any formal training/education)_, I've made changes so that if the user adds `-n` or `-u` it will add the options as exactly as  `-n` and `-u` respectively, instead of the long option. I've also added a short option `-R` as part of the `--force` option with a warning to the user not to report errors. 
I did not, however, create a conditional statement in  case the user fails to add the `--force` option when adding the **-n** and/or **-u** options nor will it check for the absence of the `**-w --workload-profile`** before adding said commands because it was not present before (but I wish I had more time to figure it out). Not sure if it's worth adding but I found these option updates extremely useful after having started a hash crack for days only to discover changeable settings to inch somewhat closer to the unbelievable hash rate reported by `--benchmark` later on and, not wanting to have to start over, it was a godsend that the changes worked somehow. It's surely saved me so much time, and hopefully it will continue to do the same for others.

_

> Time, everyone wants more of it but when we lose it, we can never get it back.

_

nucleare committed:
Updated support for -n -u --force

As of Hashcat v3.00 direct use of -n and -u was [no longer supported ](https://hashcat.net/forum/thread-5551-post-29892.html#pid29892) however, while the commands -n -u are still recognized, depending on the release version, Hashcat no longer recognize the inputs for `--gpu-loops` and `--gpu-accel` but will recognize -n and -u following a notice that the user may use '--force' to proceed with hang-tuned settings anyway. Though it's a bit sloppy (due to lack of any formal training/education), I've made changes so that if the user adds -n or -u it will add the options as -n and -u respectively, instead of the long option. I've also added a short option -R as part of the --force option with a warning to the user not to report errors. I did not, however, create a conditional statement in  case the user fails to add the `--force` option when adding the -n and/or -u options nor will it check for the absence of the `-w --workload-profile` before adding said commands simply because it was not present before. Not sure if it's worth adding but I found these extremely useful when having allowed a hash to work for a period of time before discovering the commands and not wanting to have to start over. Thank you so much for making this!

nucleare committed:
May the --force be with you

Because who doesn't wait for their long hash jobs wondering if their gpu can perform any faster only to find settings they wish they could've applied?

philsmd commented:
thanks for your contribution.

If we look at the `hashcat --help` output of newer versions of hashcat, we can see that there are still long command lang arguments for `-u` and `-u`.

```
-u, --kernel-loops
-n, --kernel-accel 
```

therefore, `-u` and `--kernel-loops` are the same (just short versus long option)
furthermore, `n` and `--kernel-accel` are the same (just short versus long option)

this means that `--gpu-loops` was replaced by `--kernel-loops` and
`--gpu-accel` was replaced by `--kernel-accel`.

We should make the same changes also for `analyze_hc_restore` . what do you think ?

Please update this pull request, if you agree.

Thank you very much.

(I **don't** think it's a completely bad idea to add `--force`, if we ensure that the user sees the warning, about NOT to use `--force` in general).

philsmd commented on analyze_hc_restore.pl:
```diff
@@ -1574,10 +1580,17 @@ sub rem_cmd_line_switch
+    
+    if ($gpu_force eq "1")
+    {
+        print "\nWARNING: adding --force to the command line overrides warnings, do not report related errors\n";
```
intentation (spaces at the start) seems wrong here

nucleare replied:
Hmm, although I was initially getting spacing errors thrown here as well, I had thought they were resolved but will revisit this.

nucleare replied:
resolved in updated commit [9f7a9e5](https://github.com/philsmd/analyze_hc_restore/pull/3/commits/9f7a9e590fd748e6f2cb50ba1d10fe3dda555c1e)

nucleare commented:
> thanks for your contribution.
> 
> If we look at the `hashcat --help` output of newer versions of hashcat, we can see that there are still long command lang arguments for `-u` and `-u`.
> 
> ```
> -u, --kernel-loops
> -n, --kernel-accel 
> ```
> 
> therefore, `-u` and `--kernel-loops` are the same (just short versus long option)
> furthermore, `n` and `--kernel-accel` are the same (just short versus long option)
> 
> this means that `--gpu-loops` was replaced by `--kernel-loops` and
> `--gpu-accel` was replaced by `--kernel-accel`.


I had actually just noticed that this morning before seeing the notification of your suggestion and am in complete agreement with you (_amidst feeling a bit embarrassed for having not noticing it earlier_). I will give it a try and test to see if it will produce the expected result. Thank you for pointing that out!

> 
> We should make the same changes also for `analyze_hc_restore` . what do you think ?
> 
> Please update this pull request, if you agree.
> 
> Thank you very much.
> 

Absolutely, will do!

> (I **don't** think it's a completely bad idea to add `--force`, if we ensure that the user sees the warning, about NOT to use `--force` in general).

Although I hadn't tested whether or not the warning would appear after modifying a .restore file and adding `-u` and/or `-n` using `analyze_hc_restore` without also adding the `--force` option, I did note that if one attempts to run hashcat with those options enabled in a new session, it won't run without the `--force` option set. Though whether or not such a message would still appear if initialing from a restore file without the `--force` option remains to be seen and tested for that matter (in case of a unexpected syntax error or something of the sort).

However, I did attempt to include such warning at line [#1590](https://github.com/nucleare/analyze_hc_restore/blob/bae030787e94281c50c3e1aa79b7ed39bdd9202d/analyze_hc_restore.pl#L1590) but merely reproduced the message hashcat would have mentioned when suggesting the necessary addition of `--force` to apply any hand tuned settings. I agree with the position of moral responsibility in ensuring users are at least privy to the frowned upon use of `--force` and may have taken, what would be considered, a more naïve approach initially, presuming that users who found a need to add such settings after the fact, would have reached by that point, a fair understanding or at least reasonable amount of frustration from trying `-w 3` or `-w 4` and ideally would've have been testing different values prior to having a need to apply it to a long running hash job (whereas the implication is that they wouldn't need to restore a session if they were still in the early stages of a hashing job) and admittedly, took the path of least resistance (AKA lazier route) by assuming they would have seen such warning multiple times by then opted out of utilizing `-w` a long time ago, considering it has been essentially deprecated and often more or less condoned for a few years now. 

However, I will admit that sort of lazy behavior is not what I would necessarily encourage nor wish to set by example and am open to suggestions for improving and/or applying a greater effort of, in how we let the user know, if by chance you have any ideas.

nucleare commented on README.md:
```diff
@@ -34,6 +34,7 @@ To list the OPTIONS available use:
 * all bug fixes are welcome
 * testing with different .restore files and/or write tests
 * solve and remove the TODOs (if any exist)
+* improve hash performance without restarting long jobs
```
long-running*

nucleare commented on analyze_hc_restore.pl:
```diff
@@ -65,6 +65,7 @@ sub usage
   print "-w | --workload-profile NUM      Set the workload profile to NUM; select between the performance profiles 1 (reduced), 2 (default) or 3 (tuned)\n";
   print "-n | --gpu-accel NUM             Set the gpu acceleration to NUM. 1, 8, 40, 80, 160\n";
   print "-u | --gpu-loops NUM             Set the gpu loops to NUM. 8 - 1024\n";
+  print "-R | --force                     Set the --force command line switch (REQUIRED if using -n -u params)\n";
```
I've changed the verbiage to better implicitly state the condoned usage

nucleare commented on analyze_hc_restore.pl:
```diff
@@ -1073,14 +1075,18 @@ sub rem_cmd_line_switch
       $optimized_kernel = "1";
     }
-    elsif (($arg eq "-n") || ($arg eq "--gpu-accel"))
+    elsif (($arg eq "-n") || ($arg eq "-n"))
```
To do: revert change and update to `--kernel-accel` to coincide with hashcat's help text

nucleare committed:
Improved verbiage consistency

-The `--gpu-accel` and `--gpu-loops` options were renamed `--kernel-accel` and `--kernel-loops` respectively, to be consistent with the help file for Hashcat. The naming scheme for the argument remains unchanged for ease of reference.
- Warning messages were added to prevent users from being able to set a `--workload-profile` while manually setting `-n` or `-u` to be consistent with Hashcat 
- The notation for the `--force` option has been updated
- Extra line spaces were removed

nucleare committed:
Minor grammatical fix.

nucleare commented:
nucleare:patch-1 has been updated and is being tested

philsmd commented on README.md:
```diff
@@ -34,6 +34,7 @@ To list the OPTIONS available use:
 * all bug fixes are welcome
 * testing with different .restore files and/or write tests
 * solve and remove the TODOs (if any exist)
+* improve hash performance without restarting long-running jobs
```
This `Hacking` section / list was actually intended to mention some points how to improve this perl script (`analyze_hc_restore.pl`).
I think your added point is actually just an use case when to use this script.
In my opinion the added point doesn't really match the goal of this section ("how we could improve this perl script"), I think we should just skip this for this patch (because it doesn't really fit with the other code changes after all).... and maybe discuss in another issue if we really need to mention how `analyze_hc_restore` can be used or what it is for (if I understand your added point correctly)

nucleare replied:
Ah thank you for pointing that out, I misinterpreted the section's intention and am glad you pointed that out. I agree with the removal and you did in fact understand correct my intention of mentioning a use case. Additionally, I thank you for your patience and mercy as you're completely right about it not having much association with the other code changes.

nucleare replied:
Resolved in commit [7e12ab7](https://github.com/philsmd/analyze_hc_restore/pull/3/commits/7e12ab736ef0047cc309fce52a14f1337e23feb8)

philsmd commented on analyze_hc_restore.pl:
```diff
@@ -63,8 +63,9 @@ sub usage
   print "-w | --workload-profile NUM      Set the workload profile to NUM; select between the performance profiles 1 (reduced), 2 (default) or 3 (tuned)\n";
-  print "-n | --gpu-accel NUM             Set the gpu acceleration to NUM. 1, 8, 40, 80, 160\n";
-  print "-u | --gpu-loops NUM             Set the gpu loops to NUM. 8 - 1024\n";
+  print "-n | --kernel-accel NUM          Set the --kernel-acceleration to NUM. 1, 8, 40, 80, 160\n";
```
maybe we should just wrtite
`Set the --kernel-accel to NUM. 1, 8, 40, 80, 160\n`
because `--kernel-acceleration` isn't really a valid command line option for hashcat... we shouldn't confuse the user

nucleare replied:
That's a great point and my apologies for not even realizing that, I shouldn't  have simply done a word replacement thinking it would be sufficient. 
Inadvertently, your mention of confusing the user made me realize some additional inconsistencies with the latest release that I feel somewhat  compelled to address but will hold off for another patch when time permits and will address just his one for now.

Just as a note for myself for future reference:
`-t` and `-T` have been disassociated from their long expression option whereas `-T` now defines `--kernel-threads` and `-t` now sets `--markov-threshold` while `--gpu-temp-disable` has been succeeded by `--hwmon-disable` and `--runtime` must be set by the long expression.

nucleare replied:
Resolved in commit [6e249ed](https://github.com/philsmd/analyze_hc_restore/commit/6e249ed27908bd07124e48656a0e1ebace266da9)

philsmd reviewed (commented):
In my opinion your changes look very good and are almost ready to be merged... I've just mentioned a few things below that I don't completely agree yet... Please have a look. thx

nucleare committed:
Provided clarity on manual workload tuning options

`--kernel-accel` more clearly identifies the setting of the outerloop step size so as not to be confused with some form of "acceleration" per se.
`--kernel-loops` was updated to identify the setting of the innerloop step more or less for aeasthetics

nucleare committed:
Removed misplaced point

Removed mention of use case previously misplaced under 'Hacking' which did not coincide with the topic of the section, which encompassed points on how to improve the script

nucleare reviewed (commented):
I believe all issues were address, please let me know if any were overlooked. The last items resolved were verbiage for kernel_accel and for the README.MD removal of unrelated use-case point

philsmd commented:
Thank you very much for your contribution, much appreciated !
